File: katas/python/Patterns/builder/builder_bad.py
# Violación del Patrón Builder: Demasiados parámetros en el constructor
# ❌ Problema: Objetos complejos con muchos parámetros opcionales

from typing import List


class Pizza:
    def __init__(
        self,
        size: str,
        crust: str,
        sauce: str,
        cheese: str,
        toppings: List[str],
        extra_cheese: bool,
        spicy_level: int,
    ):
        self.size = size
        self.crust = crust
        self.sauce = sauce
        self.cheese = cheese
        self.toppings = toppings
        self.extra_cheese = extra_cheese
        self.spicy_level = spicy_level

    def get_description(self) -> str:
        extra = ", queso extra" if self.extra_cheese else ""
        return (
            f"Pizza {self.size} con masa {self.crust}, salsa {self.sauce}, "
            f"queso {self.cheese}, ingredientes: {', '.join(self.toppings)}"
            f"{extra}, nivel picante: {self.spicy_level}/5"
        )


# ❌ Problemas al crear pizzas:
if __name__ == "__main__":
    print("=== Violación del Patrón Builder ===")

    # Difícil recordar el orden de los parámetros ❌
    margherita = Pizza(
        "grande",
        "delgada",
        "tomate",
        "mozzarella",
        ["albahaca", "tomates"],  # ❌ Debe crear array manualmente
        False,  # ❌ ¿Qué significa este booleano?
        0,  # ❌ ¿Y este número?
    )

    # Orden de parámetros confuso ❌
    carnivora = Pizza(
        "extra grande",
        "gruesa",
        "BBQ",
        "mozzarella",
        ["pepperoni", "salchicha", "tocino", "jamón"],
        True,
        3,
    )

    print("Margherita:", margherita.get_description())
    print("Carnívora:", carnivora.get_description())

File: katas/typescript/src/Patterns/builder/builder-bad.ts
// Violación del Patrón Builder: Demasiados parámetros en el constructor
// ❌ Problema: Objetos complejos con muchos parámetros opcionales

class Pizza {
  public size: string;
  public crust: string;
  public sauce: string;
  public cheese: string;
  public toppings: string[];
  public extraCheese: boolean;
  public spicyLevel: number;

  // ❌ ¡Constructor telescópico - demasiados parámetros!
  constructor(size: string, crust: string, sauce: string, cheese: string, toppings: string[], extraCheese: boolean, spicyLevel: number) {
    this.size = size;
    this.crust = crust;
    this.sauce = sauce;
    this.cheese = cheese;
    this.toppings = toppings;
    this.extraCheese = extraCheese;
    this.spicyLevel = spicyLevel;
  }

  public getDescription(): string {
    return `Pizza ${this.size} con masa ${this.crust}, salsa ${this.sauce}, queso ${this.cheese}, ingredientes: ${this.toppings.join(", ")}${this.extraCheese ? ", queso extra" : ""}, nivel picante: ${this.spicyLevel}/5`;
  }
}

// ❌ Problemas al crear pizzas:
console.log("=== Violación del Patrón Builder ===");

// Difícil recordar el orden de los parámetros ❌
const margherita = new Pizza(
  "grande",
  "delgada",
  "tomate",
  "mozzarella",
  ["albahaca", "tomates"], // ❌ Debe crear array manualmente
  false, // ❌ ¿Qué significa este booleano?
  0 // ❌ ¿Y este número?
);

// Orden de parámetros confuso ❌
const carnivora = new Pizza("extra grande", "gruesa", "BBQ", "mozzarella", ["pepperoni", "salchicha", "tocino", "jamón"], true, 3);

console.log("Margherita:", margherita.getDescription());
console.log("Carnívora:", carnivora.getDescription());

export { Pizza };
